django/db/models/expressions.py

```diff
@@ -372,77 +376,87 @@ def as_sql(self, compiler, connection):
                  (rhs_output and rhs_output.get_internal_type() == 'DurationField'))):
+            from .operators import DurationExpression
             return DurationExpression(self.lhs, self.connector, self.rhs).as_sql(compiler, connection)
-        if (lhs_output and rhs_output and self.connector == self.SUB and
-            lhs_output.get_internal_type() in {'DateField', 'DateTimeField', 'TimeField'} and
-                lhs_output.get_internal_type() == lhs_output.get_internal_type()):
-            return TemporalSubtraction(self.lhs, self.rhs).as_sql(compiler, connection)
-        expressions = []
-        expression_params = []
-        sql, params = compiler.compile(self.lhs)
-        expressions.append(sql)
-        expression_params.extend(params)
-        sql, params = compiler.compile(self.rhs)
-        expressions.append(sql)
-        expression_params.extend(params)
-        # order of precedence
-        expression_wrapper = '(%s)'
-        sql = connection.ops.combine_expression(self.connector, expressions)
-        return expression_wrapper % sql, expression_params
 
-    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
-        c = self.copy()
-        c.is_summary = summarize
-        c.lhs = c.lhs.resolve_expression(query, allow_joins, reuse, summarize, for_save)
-        c.rhs = c.rhs.resolve_expression(query, allow_joins, reuse, summarize, for_save)
-        return c
+        if lhs_output or rhs_output:
+            dispatchers = (self.expression_dispatchers.get(self.connector),
+                           self.expression_dispatchers.get(CombinedExpression.ANY))
 
+            for dispatcher in filter(bool, dispatchers):
+                try:
+                    klass = dispatcher(lhs_output or fields.Field(), rhs_output or fields.Field())
 
-class DurationExpression(CombinedExpression):
-    def compile(self, side, compiler, connection):
-        if not isinstance(side, DurationValue):
-            try:
-                output = side.output_field
-            except FieldError:
-                pass
-            else:
-                if output.get_internal_type() == 'DurationField':
-                    sql, params = compiler.compile(side)
-                    return connection.ops.format_for_duration_arithmetic(sql), params
-        return compiler.compile(side)
+                    # avoid recursion in case CombinedExpression maps to another
+                    # CombinedExpression, but with a different connector
+                    if type(self) != klass:
+                        try:
+                            output_field = self._output_field_or_none
+                        except FieldError:
+                            output_field = None
+
+                        expression = klass(self.lhs, self.connector, self.rhs, output_field=output_field)
+                        return compiler.compile(expression)
```

Comment:
Keep the `try` body to a minimum by only including the code that could raise the expected `NotImplementedError` exception. Use an `else` branch if necessary.